### PR TITLE
APP-85-native-simple-sign

### PR DESCRIPTION
**Pull request by tarikstafford**

### CODE REVIEW

**Comment by sampettersson** [comment]
@tarikstafford tried running this locally but it Ugglan doesn't compile 😢

File: Projects/hCoreUI/Sources/ButtonRow.swift [hunk 1]
```diff
@@ -70,12 +70,14 @@ public struct ButtonRowViewWrapper {
     public let onTapSignal: Signal<Void>
     public let type: ButtonType
     public let animate: Bool
-    public let isEnabled: Bool
+    private let isEnabled: Bool
```
Comment (sampettersson) [inline Projects/hCoreUI/Sources/ButtonRow.swift:73]: I guess remove this if we use it as a signal instead?

File: Projects/App/Sources/Generic/AppFlow.swift [hunk 2]
```diff
@@ -6,27 +6,34 @@ import hCoreUI
...
     let window: UIWindow
+    let bag = DisposeBag()
 
     init(window: UIWindow) {
         self.window = window
         self.window.rootViewController = rootNavigationController
     }
+
+    func presentLoggedIn() {
+        ApplicationState.preserveState(.loggedIn)
```
Comment (sampettersson) [inline Projects/App/Sources/Generic/AppFlow.swift:21]: Isn't this done inside `LoggedIn`?

File: Projects/App/Sources/Generic/AppFlow.swift [hunk 3]
```diff
@@ -95,9 +102,11 @@ extension EmbarkOnboardingFlow {
         case .login:
             return Presentation(
                 Login(),
-                style: .detented(.medium, .large),
-                options: [.allowSwipeDismissAlways, .defaults]
-            )
+                style: .detented(.large),
```
Comment (sampettersson) [inline Projects/App/Sources/Generic/AppFlow.swift:105]: it should be `.detented(.medium)` for Sweden. I guess we should use `preferredPresentationStyle` and just `.default`.

**Review by sampettersson — CHANGES_REQUESTED** [review]

File: Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        let form = FormView()
+
+        let masking = Localization.Locale.currentLocale.market.masking
+
+        viewController.title = L10n.bankidLoginTitle
+
+        let label = UILabel()
+        label.text = Localization.Locale.currentLocale.market.labelTitle
+
+        let textField = UITextField()
```
Comment (sampettersson) [inline Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:29]: What do you think about making a utility func on masking that applies it to a `UITextField`, something like `Masking.applyTo(_ textField: UITextField) -> Bag` which handles `isValid`, `textContentType`, `keyboardType`, etc.

Comment (tarikstafford) [reply Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:29]: How would you add `isValid` in this case? Perhaps it needs to be thought out a bit more before we add a convenience function for it.

Comment (tarikstafford) [reply Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:29]: I guess `maskValue` would make sense, although I am not really using this. Actually it's never really masking the value is it? Just formatting it.

Comment (tarikstafford) [reply Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:29]: I tried doing this a few ways, but I don't really like the idea of a method returning a signal for validity in a masking object. Do you think we should split up `masking` and `validation` -> perhaps we can change `maskType` to something like `textFieldType` or `textFieldContentType` ?

Comment (tarikstafford) [reply Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:29]: Also the textField settings would be configurable from the type itself perhaps? Then masking would just have an `applyMask` and `validation` would have a `isValidSignal`

Comment (sampettersson) [reply Projects/App/Sources/Screens/BankIDLoginScreen/SimpleSignLogin.swift:38]: @tarikstafford don't we always wanna mask? it makes sure that the user can only enter the valid characters, also it inserts the hyphens at the correct places and so forth also we have the same predictable behaviour in all places of the app. And then we can just use `unmaskValue` before sending to backend.

File: Projects/hCore/Sources/Masking.swift [hunk 5]
```diff
@@ -16,10 +17,34 @@ public enum MaskType: String {
...
+    public func isValidSignal(_ textField: UITextField) -> ReadSignal<Bool> {
+        textField.distinct().map { text in isValid(text: text) }
+    }
+
+    public func applyMasking(_ textField: UITextField) -> Disposable {
+        let bag = DisposeBag()
+
+        bag += textField.distinct().onValue { text in
+            let newValue = maskValue(text: text, previousText: previousText)
+            $previousText.value = newValue
```
Comment (sampettersson) [inline Projects/hCore/Sources/Masking.swift:41]: Way nicer than before! 😄

**Review by sampettersson — APPROVED** [review]

**Review by sampettersson — APPROVED** [review]

**Pull request merged**